app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,388 @@
+import org.kiwix.kiwixmobile.utils.ServerUtils;
+import org.kiwix.kiwixmobile.wifi_hotspot.HotspotService;
+import org.kiwix.kiwixmobile.zim_manager.fileselect_view.SelectionMode;
+import org.kiwix.kiwixmobile.zim_manager.fileselect_view.adapter.BookOnDiskDelegate;
+import org.kiwix.kiwixmobile.zim_manager.fileselect_view.adapter.BooksOnDiskAdapter;
+import org.kiwix.kiwixmobile.zim_manager.fileselect_view.adapter.BooksOnDiskListItem;
+
+import static org.kiwix.kiwixmobile.wifi_hotspot.HotspotService.ACTION_LOCATION_ACCESS_GRANTED;
+import static org.kiwix.kiwixmobile.wifi_hotspot.HotspotService.ACTION_START_SERVER;
+import static org.kiwix.kiwixmobile.wifi_hotspot.HotspotService.ACTION_STOP_SERVER;
+import static org.kiwix.kiwixmobile.wifi_hotspot.HotspotService.ACTION_TOGGLE_HOTSPOT;
+
+public class ZimHostActivity extends BaseActivity implements
+    ZimHostCallbacks, ZimHostContract.View, LocationCallbacks {
+
+  @BindView(R.id.startServerButton)
+  Button startServerButton;
+  @BindView(R.id.server_textView)
+  TextView serverTextView;
+  @BindView(R.id.recycler_view_zim_host)
+  RecyclerView recyclerViewZimHost;
+
+  @Inject
+  ZimHostContract.Presenter presenter;
+
+  @Inject
+  AlertDialogShower alertDialogShower;
+
+  @Inject
+  LocationServicesHelper locationServicesHelper;
+
+  private static final String TAG = "ZimHostActivity";
+  private static final int MY_PERMISSIONS_ACCESS_FINE_LOCATION = 102;
+  public static final String SELECTED_ZIM_PATHS_KEY = "selected_zim_paths";
+  private static final String IP_STATE_KEY = "ip_state_key";
+  private ProgressDialog progressDialog;
+
+  private BooksOnDiskAdapter booksAdapter;
+  private BookOnDiskDelegate.BookDelegate bookDelegate;
+  private HotspotService hotspotService;
+  private String ip;
+  private ServiceConnection serviceConnection;
+
+  @Override
+  protected void onCreate(@Nullable Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_zim_host);
+
+    setUpToolbar();
+
+    if (savedInstanceState != null) {
+      ip = savedInstanceState.getString(IP_STATE_KEY);
+      layoutServerStarted();
+    }
+    bookDelegate =
+        new BookOnDiskDelegate.BookDelegate(sharedPreferenceUtil,
+            null,
+            null,
+            bookOnDiskItem -> {
+              select(bookOnDiskItem);
+              return Unit.INSTANCE;
+            });
+    bookDelegate.setSelectionMode(SelectionMode.MULTI);
+    booksAdapter = new BooksOnDiskAdapter(bookDelegate,
+        BookOnDiskDelegate.LanguageDelegate.INSTANCE
+    );
+
+    presenter.attachView(this);
+
+    presenter.loadBooks();
+    recyclerViewZimHost.setAdapter(booksAdapter);
+
+    serviceConnection = new ServiceConnection() {
+
+      @Override
+      public void onServiceConnected(ComponentName className, IBinder service) {
+        HotspotService.HotspotBinder binder = (HotspotService.HotspotBinder) service;
+        hotspotService = binder.getService();
```

Comment:
this doesn't need to be a field anymore, it could infact be 1 line but that might look ugly
```
((HotspotService.HotspotBinder) service).getService().registersterCallback(ZimHostActivity.this)
```
you can use `ctrl+alt+v` to extract a variable of the binder if it looks too ugly in IDE

Reply:
I believe this looks good now.  f54791deb55cbc2f4beed1b1d5b1901b5e2b6537

Reply:
>this doesn't need to be a field anymore

Was there a problem removing it as a field?

Reply:
We need to null check `hotspotService` before disconnecting the service